docs/blog/_posts/2016-12-05-implicit-function-types.md

```diff
@@ -0,0 +1,170 @@
+The inputs are typically what a computation can access in its
+environment. Interesting tasks that are by nature comonadic are
+
+ - passing configuration data to the parts of a system that need them,
+ - managing capabilities for security critical tasks,
+ - wiring components up with dependency injection,
+ - defining the meanings of operations with type classes,
+ - more generally, passing any sort of context to a computation.
+
+Implicit function types are a suprisingly simple and general way to
+make coding patterns solving these tasks abstractable, reducing
+boilerplate code and increasing applicability.
+
+*First Step* My pull request is first implementation. In solves the
+ problem in principle, but it introduces some run-time overhead. The
+ next step will be to eliminate the run-time overhead through some
+ simple optimizations.
+
+
+## Comparison with Monads
+
+One can use monads for these tasks, and some people do. For instance
+the `Reader` monad is used to abstract over accessing one entry in the
+environment.  But the code for doing so quickly becomes complex and
+inefficient, in particular when combining several contextual
+accesses. Monads don't compose in general, and therefore even simple
+combinations need to be expressed on the level of monad transformers,
+at the price of much boilerplate and complexity. Recognizing this,
+peaple have recently experimented with free monads, which alleviate
+the composibility problem, but at the price of introducing a whole new
+level of interpretation.
+
+## Implicit Parameters
+
+In a functional setting, the inputs to a computation are most
+naturally expressed as _parameters_. One could simply augment
+functions to take additional parameters that represent configurations,
+capabilities, dictionaries, or whatever contextual data the functions
+need. The only downside with this is that often there's a large
+distance in the call graph between the definition of a contextual
+element and the site where it is used. Conseuqently, it becomes
+tedious to define all those intermediate parameters and to pass them
+along to where they are eventually consumed.
+
+Implicit parameters solve one half of the problem. Implicit
+parameters do not have to be propagated using boilerplate code; the
+compiler takes care of that. This makes them practical in many
+scenarios where plain parameters would be too cumbersome. For
+instance, type classes would be a lot less popular if one would have
+to pass all dictionaries by hand. Implicit parameters are also very
+useful as a general context passing mechanism. For instance in the
+_dotty_ compiler, almost every function takes an implicit context
+parameter which defines all elements relating to the current state of
+the compilation. This is in my experience much better than the cake
+pattern because it is lightweight and can express context changes in a
+purely functional way.
+
+The main downside of implicit parameters is the verbosity of their
```

Comment:
> But that's nothing new. All free bindings of a piece of code affect its semantics. So the semantics of any expression that's not closed depends on its location (?)

Implicits simply hide the free variables in use, so that figuring out the context requires typechecking rather than just parsing. So, after you look up `implicitly`'s type, you discover that `implicitly[Foo]` is an open term (which is sugar for open term `implicitly[Foo](theImplicitDefinition)`). This is confusing enough it was (indirectly) discussed above: https://github.com/lampepfl/dotty/pull/1775#discussion_r91242465